common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+class DirRef::Entry {
+ public:
+  // The name of the entry.
+  //
+  // This is exposed as a null-terminated C-string as that is the most common
+  // representation.
+  auto name() const -> const char* { return dent_->d_name; }
+
+  // Test whether the entry is known to be a directory.
+  //
+  // An empty optional indicates it is unknown what kind of entry this is, while
+  // an engaged optional contains a bool indicating whether it is known a
+  // directory or known a non-directory entry.
+  auto is_dir() const -> std::optional<bool> {
+    return dent_->d_type == DT_UNKNOWN ? std::nullopt
+                                       : std::optional(dent_->d_type == DT_DIR);
+  }
+
+ private:
+  friend Dir::Reader;
+  friend Dir::Iterator;
+
+  Entry() = default;
+  explicit Entry(dirent* dent) : dent_(dent) {}
+
+  dirent* dent_ = nullptr;
+};
+
+// An iterator into a `DirRef::Reader`, used for walking the entries in a
+// directory.
+//
+// Most of the work of iterating a directory is done when constructing the
+// `Reader`, when constructing the beginning iterator, or when incrementing the
+// iterator.
+class DirRef::Iterator
+    : public llvm::iterator_facade_base<Iterator, std::input_iterator_tag,
+                                        const Entry> {
+ public:
+  // Default construct a general end iterator.
+  Iterator() = default;
+
+  auto operator==(const Iterator& rhs) const -> bool {
+    CARBON_DCHECK(dirp_ == nullptr || rhs.dirp_ == nullptr ||
+                  dirp_ == rhs.dirp_);
+    return entry_.dent_ == rhs.entry_.dent_;
+  }
+
+  auto operator*() const -> const Entry& { return entry_; }
+  auto operator++() -> Iterator&;
+
+ private:
+  friend Dir::Reader;
+
+  // Construct a begin iterator for a specific directory stream.
+  explicit Iterator(DIR* dirp) : dirp_(dirp) {
+    // Increment immediately to populate the initial entry.
+    ++*this;
+  }
+
+  DIR* dirp_ = nullptr;
+  Entry entry_;
+};
+
+// A reader for a directory.
+//
+// This class owns a handle to a directory that is set up for reading the
+// entries within the directory. Because it owns a handle to the directory, it
+// also implements the full `DirRef` API for convenience.
+//
+// Beyond the `DirRef` API, this object can be iterated as a range to visit all
+// the entries in the directory.
+//
+// Note that it is unspecified whether entries added or removed prior to being
+// visited while iterating. Iterating also cannot be re-started once begun --
+// this models an input iterable range, not even a forward iterable range.
+//
+// This type allows intentional "slicing" to the `DirRef` base class as that is
+// a correct and safe conversion to pass a non-owning reference to a directory
+// to another function, much like binding a reference to an owning type is
+// implicit.
+class DirRef::Reader : public DirRef {
+ public:
+  Reader() = default;
+  Reader(Reader&& arg) noexcept : DirRef(arg.dfd_), dirp_(arg.dirp_) {
+    arg.dirp_ = nullptr;
+  }
```

Comment:
```suggestion
  Reader(Reader&& arg) noexcept : DirRef(std::exchange(arg.dfd_, -1)), dirp_(std::exchange(arg.dirp_, nullptr)) {}
```

`Dir` moves the `dfd_` so `Reader` should do the same, right?

Reply:
Technically, its just `dirp_` that has ownership here, the `dfd_` is essentially just a cache of `dirfd(dirp_)` to read the fd back out of thu underlying directory struct.

But no reason to leave `dfd_` in a non-`-1` state after moving out of the arg. That said, doing it without `std::exchange` pending discussion of that.